### ISSUE TITLE
StanFit4Model.extract slow for recovering generated quantities

### ISSUE THREAD

**Issue by anguswilliams91** [stan-dev/pystan#462]
#### Summary:
Recovering data from `generated quantities` using `StanFit4Model.extract` is slow.

#### Description:
Using the example code 

```
data {
    int<lower=0> n;
}
generated quantities {
    vector[n] theta;
    for (j in 1:n) theta[j] = normal_rng(0., 1.);
}
```

I find that `fit.extract()` is very slow with increasing `n`. 
I had a discussion about this [on the forum](http://discourse.mc-stan.org/t/fit-extract-takes-a-long-time/4159).
There was a suggestion to use `fit.extract('theta', permuted=False)` instead, but I could not see any different in performance when I used this.
I found a hacky approach to be several orders of magnitude faster : `[float(i) for i in fit.sim['samples'][0].chains.values()][:-1]`, so it seems in principle that the extract of the generated quantities could be sped up.
On the other hand, I may be missing something or misunderstanding the intended use of these methods - apologies if that is the case.
There is a graph that shows the timings on my machine with increasing `n` in the readme of [this repo](https://github.com/anguswilliams91/pystan_timing).

#### Reproducible Steps:
In the [previously mentioned repo](https://github.com/anguswilliams91/pystan_timing) there is the code that I used to generate the graph, and an `environment.yml` to reproduce the anaconda environment I was using.
To produce the plot, create the anaconda environment and then run the python script `stan_timing.py`.
Since this is a timing thing, the results won't perfectly replicate.
Apologies if there is a silly mistake anywhere that renders this issue redundant!

#### PyStan Version:
2.17.1.0

#### Python Version:
3.6.5

#### Operating System:
macOS High Sierra

**Comment by ariddell** [comment]
Thanks for this report. I think your test case, in particular, will be valuable no matter what we do. We should track the speed at which one is able to extract 100,000 parameters using fixed_param and a single (or a very small number) iteration and a single chain. It should not change much from release to release.

**Comment by ahartikainen** [comment]
Hi,
the issue is the following loop:

    extracted_par = []
    for n in tidx[par]:
        chains = pystan.misc._get_samples(n, fit.sim, False)
        n_save = fit.sim['n_save'][0]
        if not False:
            n_save = n_save - fit.sim['warmup2'][0]
        samples = np.column_stack(chains)
        extracted_par.append(samples[:, :, np.newaxis])

And in there it is the `chains = pystan.misc._get_samples(n, fit.sim, False)` line. (it took me a 800 us for one call)

`pystan.misc._get_samples` then calls `pystan._misc.get_samples` which is a cython function.

    @cython.boundscheck(False)
    @cython.wraparound(False)
    cpdef get_samples(int n, dict sim, inc_warmup):
        """See documentation in misc.py"""
        cdef int i
        cdef double[:] s
        cdef int num_chains = sim['chains']
        if sim['warmup'] == 0:
            inc_warmup = True

        # the following assumes each chain has same length, same warmup
        cdef int num_warmup = sim['warmup2'][0]
    
        nth_key = list(sim['samples'][0]['chains'].keys())[n]
        ss = []
        for i in range(num_chains):
            s = sim['samples'][i]['chains'][nth_key]
            if inc_warmup:
                ss.append(s)
            else:
                ss.append(s[num_warmup:])
        return ss

So in every iteration this function is called. This is not a problem in the common use case where the data is not that wide, but deep (few parameters, many samples), but it becomes a problem when the your data is opposite of that (many parameters, few (one) samples).

There could be a possibility to streamline our cython - python interface. E.g. save parameters directly to ndarrays (theta.shape = [samples, *shape]) instead of odict ({'theta[0]' : arr, 'theta[1]' : arr}). Not sure when this is going to happen (PyStan 3 ?).

**Comment by ariddell** [comment]
This is really helpful. We definitely should use something other than a
dict. (I do, however, like the clarity of code when using the dict.)
Maybe xarray would provide something that works better.

My goal is to get something resembling feature parity between PyStan 3
and the current Pystan by the end of the first week in June. This sort
of thing should be resolved by then.

**Comment by anguswilliams91** [comment]
Thanks @ahartikainen for the response (both here and on the forum). I think it would make sense to extract the parameters directly into `ndarray`s.  Looking forward to Pystan 3!

**Comment by Bonnevie** [comment]
Is there any progress on this issue?
At the moment, using `to_dataframe` and `extract` on a matrix completion variable of shape 310 x 177 takes so long that it renders the function almost unusable. Of course operations cannot be sped up arbitrarily, but by accessing the `sim['samples']` property as done by @anguswilliams91, I can perform the same procedure manually almost instantaneously.
As before, profiling points to `get_samples`, but I cannot see why it would be so much more costly. Is it the repeated calls to `list()` and `.keys()` in `nth_key`? Some preliminary testing seems to indicate that this might be the issue: `get_samples` is timed to 24 ms and the `nth_key` line is timed independently to 23.8 ms, with the `keys` command alone only contributing 138 ns.

I would be happy to help with this - at the moment it makes pystan almost unusable with large `transformed parameters` or `generated quantities`.

**Comment by riddell-stan** [comment]
The plan for the next version is to store draws internally in an array 
with shape: (num_params, num_draws, num_chains) (see 
https://github.com/stan-dev/stan/wiki/User-Interface-Guidelines-for-Developers 
)

This should make some operations quite a bit faster. I suspect it will 
solve your problem.

There should be a pre-alpha release of PyStan 3 in a couple of weeks. It 
would be great to see if we can get this particular problem solved (and 
tested).

**Comment by ariddell** [comment]
It's coming from RStan 3. There's a good reason for it, but I don't 
recall it right now. It might have something to do with making adding 
chains or draws easier or faster.

**Comment by ahartikainen** [comment]
It probably follows Fortran order.

**Comment by Bonnevie** [comment]
A possible fix for the meantime if I diagnosed the issue correctly:

the potentially problematic line (i.e. if `keys` produces a generator with 30000 elements and a list then has to be instantiated with all those elements 30000 times over) is,
```
nth_key = list(sim['samples'][0]['chains'].keys())[n]
```
seems to just be a convoluted way of getting the name of the n'th variable, although disconcertingly by accessing an unordered object (although I have never quite figured out when or indeed if the keys generator changes order). These names are the `flatnames` used by the model, and are saved in `sim['fnames_oi']`. By simply accessing that object instead, we avoid constructing the lists. I inspected the objects, and `nth_key` and `sim['fnames_oi']` do indeed seem to match, so one could be swapped for the other.

**Comment by ahartikainen** [comment]
Common interface with PyStan 3 is more important.


That said, if anyone is up to do some serious numpy magic, then we probably could update our extract method.

E.g. `operator.itemgetter` + numpy reshaping could speed up process. This is usually harder when data is nD.

Maybe we can create a test set with wide, long and wide-long data and see possible speeds.

**Comment by Bonnevie** [comment]
@ahartikainen was it not possible to correct the single line, or did I misunderstand you? I could do it if that's easier, just seems like a rather minor change for a PR.

Reshaping of slices of the (num_params, num_draws, num_chains) matrix might work in reconstructing ND variables, although it does require that `num_params` is appropriately ordered.

**Comment by riddell-stan** [comment]
I'd wager I wrote this code and was copying the RStan code. A fix would 
be welcome!

The major thing to watch out for is how the multidimensional arrays are 
flattened and unflattened. Numpy, by default, uses C order (row major), 
whereas Stan uses Fortran order (column major).

**Comment by ahartikainen** [comment]
@Bonnevie

Raw pandas DataFrame with all parameters is fast with just extracting stuff from original dictionary (`pandas.DataFrame.from_dict` + concat chains).

I'm thinking of updating extract method.

**Comment by ahartikainen** [comment]
This is fast pandas

    pd.DataFrame.from_dict(fit.sim['samples'][0].chains)

And then concat for many chains.

**Comment by ahartikainen** [comment]
I made an PR #526 and can also update `to_dataframe` function later.

**Comment by ahartikainen** [comment]
For really large models, `check_hmc_diagnostics` takes a long time. Not sure if we should make it throw a warning for large models?

Also a new to_dataframe is now implemented in the PR.

**Comment by riddell-stan** [comment]
What does RStan do? I can't recall if they even run the check automatically.

We could do many things. For example, we could skip it if we detect the 
model is enormous. If we skip it, we could emit a warning and tell 
people to run the check manually.

**Comment by ahartikainen** [comment]
here are some comparisons with the new function.

https://github.com/ahartikainen/misc_bayes/tree/master/PyStan_extract

In these examples `check_hmc_diagnostics=False`, even when image titles do not contain it.

**Comment by Bonnevie** [comment]
@ahartikainen that work is a real life-saver! Many of the models I have been toying with have matrix-valued parameters and the down-time has been pretty discouraging previously.

Do you think it likely that the diagnostics can be sped up in the same manner? I assume part of their slow-down was due to calling `extract` any way?

**Comment by bob-carpenter** [comment]
Autocorrelation isn't so bad because it can be done with an FFT.  If that's not built-in somewhere in Python, you can use the C++ in stan-dev/stan for the calculations.

**Comment by ahartikainen** [comment]
Sure, fft is super fast compared to naïve method, but for 100k times it still might be slow

We use Stan-math autocov in here https://github.com/stan-dev/pystan/blob/develop/pystan/_chains.pyx

I checked the `hmc_diagnostics` and noticed it calls `fit.summary` multiple times. We could cache that call, call special functions only or something else. I will explire the options.

**Comment by ahartikainen** [comment]
Fixed by #526 & #532 & #540
